File: 279.perfect-squares.py
#
# @lc app=leetcode id=279 lang=python
#
# [279] Perfect Squares
#

# @lc code=start
class Solution(object):
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        # BFS
        # squares = [i**2 for i in range(1, int(n**0.5)+1)]
        # d, q, nq = 1, {n}, set()
        # while q:
        #     for node in q:
        #         for square in squares:
        #             if node == square: return d
        #             if node < square: break
        #             nq.add(node-square)
        #     q, nq, d = nq, set(), d+1

        dp = [0] + [float('inf')]*n
        for i in range(1, n+1):
            dp[i] = min(dp[i-j*j] for j in range(1, int(i**0.5)+1)) + 1
        return dp[n]
        
# @lc code=end

File: 279.perfect-squares.cpp
/*
 * @lc app=leetcode id=279 lang=cpp
 *
 * [279] Perfect Squares
 */
#include <queue>
#include <vector>
using namespace std;

// @lc code=start
class Solution {
public:
    int numSquares(int n) {
        // 平方数据
        vector<int> squares;
        int square = 1, diff = 3;
        while(square <= n) {
            squares.push_back(square);
            square += diff;
            diff += 2;
        }
        queue<int> q;
        vector<bool> marked(n+1, false);
        q.push(n);
        marked[n] = true;
        int level = 0;
        while (!q.empty()) {
            int size = q.size();
            level++;
            while (size-- > 0) {
                int cur = q.front();
                q.pop();
                for (int s: squares) {
                    int next = cur - s;
                    if (next < 0)
                        break;
                    if (next == 0)
                        return level;
                    if (marked[next])
                        continue;
                    marked[next] = true;
                    q.push(next);
                }
            }
        }
        return n;
    }
};
// @lc code=end

